test/dotnet/input/InsertBulk/insertBulkErrors.txt

```diff
@@ -555,4 +555,77 @@ drop table sourceTable
 drop table sourceTable1
 drop table destinationTable
 
- SET implicit_transactions OFF
\ No newline at end of file
+ SET implicit_transactions OFF
+
+ # identity
+create table sourceTable(c1 int identity PRIMARY KEY, c2 CHAR(1024))
+ set IDENTITY_INSERT sourceTable ON
+INSERT INTO sourceTable (c1, c2) SELECT generate_series(1, 1001, 1), 'Foo';
+ set IDENTITY_INSERT sourceTable OFF
+create table destinationTable(c1 int identity PRIMARY KEY, c2 CHAR(1024))
+ set IDENTITY_INSERT destinationTable ON
+INSERT INTO destinationTable (c1, c2) VALUES(1001, 'Foo')
+ set IDENTITY_INSERT destinationTable OFF
+keepidentityinsertbulk#!#sourceTable#!#destinationTable
+create table sourceTable1(c1 int identity PRIMARY KEY, c2 CHAR(1024))
+ set IDENTITY_INSERT sourceTable1 ON
+INSERT INTO sourceTable1 (c1, c2) VALUES(1001, 'Foo')
+INSERT INTO sourceTable1 (c1, c2) SELECT generate_series(1, 1000, 1), 'Foo';
+ set IDENTITY_INSERT sourceTable1 OFF
+keepidentityinsertbulk#!#sourceTable1#!#destinationTable
+Select count(c1) from sourceTable
+select count(c1) from sourceTable1
+select count(c1) from destinationTable
+select Ident_CURRENT('sourceTable')
+select Ident_CURRENT('sourceTable1')
+select Ident_CURRENT('destinationTable')
+drop table sourceTable
+drop table sourceTable1
+drop table destinationTable
+
+create table sourceTable(c1 int identity, c2 CHAR(1024), check (c1 < 1000))
+ set IDENTITY_INSERT sourceTable ON
+INSERT INTO sourceTable (c1, c2) SELECT generate_series(1, 1001, 1), 'Foo';
+ set IDENTITY_INSERT sourceTable OFF
+create table destinationTable(c1 int identity, c2 CHAR(1024), check (c1 < 500))
+ set IDENTITY_INSERT destinationTable ON
+INSERT INTO destinationTable (c1, c2) VALUES(1001, 'Foo')
```

Comment:
Use transinsertbulk since we enforce check constraints.

What is the behaviour of TSQL in case of check constraints off and KeepIdentity is on, I am guessing user can go on inserting c1>1000. 

Can we check that? These tests dont really check that

Reply:
yes, in TSQL if check constraint is off and KeepIdentity is on, we can insert c1>=500. But we can't really check that in babelfish as Check constraint is enabled by default and we can't disable that using ALTER TABLE for new rows being inserted as 

In PostgreSQL, when you execute ALTER TABLE destinationTable NOCHECK CONSTRAINT ALL;, it only skips the validation of existing rows against the check constraints. However, it does not disable the check constraints for new rows being inserted or updated.

So, when you try to insert a new row with c1 = 501, which violates the check constraint c1 < 500, PostgreSQL still enforces the check constraint and raises the error new row for relation "destinationtable" violates check constraint "destinationtable_c1_check".

The NOCHECK CONSTRAINT option in PostgreSQL is primarily used when you need to load data that might violate the constraints temporarily, and then you can validate and clean up the data later. However, it does not allow you to bypass the constraints for new data being inserted or updated.

sample query:
```
1> SELECT set_config('babelfishpg_tsql.escape_hatch_nocheck_existing_constraint', 'ignore', 'false')
2> go
set_config                                                                                                                                                                                                                                                      
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
ignore                                                                                                                                                                                                                                                          

(1 rows affected)
1> create table destinationTable(c1 int identity, c2 CHAR(1024), check (c1 < 500))
2> go
1> ALTER TABLE destinationTable NOCHECK CONSTRAINT ALL;
2> go
1> set identity_insert destinationTable ON;
2> go
1> insert into destinationTable (c1, c2) values (501, 'qd')
2> go
Msg 547, Level 16, State 1, Server BABELFISH, Line 1
new row for relation "destinationtable" violates check constraint "destinationtable_c1_check"
```